proposed/http-client/http-client.md

```diff
@@ -21,9 +21,9 @@ implementations. This would make libraries more stable since the reduced number
 dependencies and the likelihood to get in version conflicts is reduced.
 
 The second goal is that all HTTP clients should follow the [Liskov substitutions principle][Liskov].
-This means that all clients should act the same when sending a request. By default a HTTP client
-should not follow redirect nor throw exceptions on HTTP responses with status 4xx or 5xx.
-
+This means that all clients should act the same when sending a request. A consuming
+library SHOULD expect a HTTP client to NOT follow redirects NOR throw exceptions based
+on HTTP response statuses.
```

Comment:
Because this trade-off is ugly either way, I'm leaning towards suggesting a second argument be added, e.g. signature as follows:

    sendRequest(RequestInterface $request, bool $follow = true): ResponseInterface

This way, the default behavior is appropriate for the majority of use-cases - and for rare cases where the client needs to know if there was an intermediary redirect, they can set `$follow` to `false`.

I'm afraid there's no other way to satisfy the most common use-case without precluding the other less-common use-cases.

If we have to manually implement following redirects in almost every piece of client-code, anything that depends on this abstraction is going to be error-prone and will need tests and mocks for redirect handling, which would be pretty awkward.

If you're opposed to having this as an option, I'd say clients *should* follow redirects, and consumers will need to use a different client for those less-common cases where they need to know if there's a redirect. The majority of clients probably don't need to know, e.g. those fetching pages, data or assets/images - likely only special clients like spiders and diagnostic tools would need to know if there's a redirect.

What doesn't make sense, is to design for the minority use-case.